be/src/vec/core/sort_cursor.h

```diff
@@ -314,4 +340,224 @@ struct MergeSortBlockCursor {
+                ++next_child_idx;
+            }
+        }
+
+        return next_child_idx;
+    }
+
+    /// This is adapted version of the function __sift_down from libc++.
+    /// Why cannot simply use std::priority_queue?
+    /// - because it doesn't support updating the top element and requires pop and push instead.
+    /// Also look at "Boost.Heap" library.
+    void ALWAYS_INLINE update_top(bool check_in_order) {
+        size_t size = queue.size();
+        if (size < 2) {
+            return;
+        }
+
+        auto begin = queue.begin();
+
+        size_t child_idx = next_child_index();
+        auto child_it = begin + child_idx;
+
+        /// Check if we are in order.
+        if (check_in_order && (*child_it).greater(*begin)) {
+            if constexpr (strategy == SortingQueueStrategy::Batch) {
+                update_batch_size();
+            }
+            return;
+        }
+
+        next_child_idx = 0;
+
+        auto curr_it = begin;
+        auto top(std::move(*begin));
+        do {
+            /// We are not in heap-order, swap the parent with it's largest child.
+            *curr_it = std::move(*child_it);
+            curr_it = child_it;
+
+            // recompute the child based off of the updated parent
+            child_idx = 2 * child_idx + 1;
+
+            if (child_idx >= size) {
+                break;
+            }
+
+            child_it = begin + child_idx;
+
+            if ((child_idx + 1) < size && (*child_it).greater(*(child_it + 1))) {
+                /// Right child exists and is greater than left child.
+                ++child_it;
+                ++child_idx;
+            }
+
+            /// Check if we are in order.
+        } while (!((*child_it).greater(top)));
+        *curr_it = std::move(top);
+
+        if constexpr (strategy == SortingQueueStrategy::Batch) {
+            update_batch_size();
+        }
+    }
+
+    /// Update batch size of elements that client can extract from current cursor
+    void update_batch_size() {
+        assert(!queue.empty());
+
+        auto& begin_cursor = *queue.begin();
+        size_t min_cursor_size = begin_cursor->get_size();
+        size_t min_cursor_pos = begin_cursor->pos;
+
+        if (queue.size() == 1) {
+            batch_size = min_cursor_size - min_cursor_pos;
+            return;
+        }
+
+        batch_size = 1;
+        size_t child_idx = next_child_index();
+        auto& next_child_cursor = *(queue.begin() + child_idx);
+        if (min_cursor_pos + batch_size < min_cursor_size &&
+            next_child_cursor.greater_with_offset(begin_cursor, 0, batch_size)) {
+            ++batch_size;
+        } else {
+            return;
+        }
+        if (begin_cursor.totally_less_or_equals(next_child_cursor)) {
+            batch_size = min_cursor_size - min_cursor_pos;
+            return;
+        }
+
+        while (min_cursor_pos + batch_size < min_cursor_size &&
```

Comment:
the best way here maybe is binary search ?

Reply:
binary search will introduce least log(batch_size) comparison